### PR TITLE
Improve error handling for http to gRPC invocation

### PR DESCRIPTION
**Pull request by youngbupark**
# Description

Improve error handling for http to gRPC invocation

### HTTP to gRPC

1. Headers: No change
2. Response No change
3. gRPC Response body conversion
   * gRPC callee returns `OK` status
     * user app returns Dapr [common.InvokeResponse](https://github.com/dapr/dapr/blob/1c21def42063c0f3a58ca19b69dcd46874ad3611/dapr/proto/common/v1/common.proto#L70-L80) protobuf message.
       * content_type in [common.InvokeResponse](https://github.com/dapr/dapr/blob/1c21def42063c0f3a58ca19b69dcd46874ad3611/dapr/proto/common/v1/common.proto#L70-L80) to HTTP contentType header
       * data in [common.InvokeResponse](https://github.com/dapr/dapr/blob/1c21def42063c0f3a58ca19b69dcd46874ad3611/dapr/proto/common/v1/common.proto#L70-L80) to HTTP response body
     * user app returns custom protobuf message
       * Populate `application/x-protobuf` to HTTP ContentType header
       * Populate the serialized protobuf message (bytes array) to HTTP response body
   * gRPC callee returns `non-OK` status
     * Populate `application/json` to HTTP ContentType Header
     * Using [jsonpb](https://godoc.org/github.com/golang/protobuf/jsonpb), Json-serialize gRPC rich error model [Status](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto) protobuf message to HTTP response body.

example

```json
{
   "code": 14,
   "message": "Unavailable",
   "details": [{
      "@type": "type.googleapis.com/google.rpc.RetryInfo",
      ...
    }]
}
```

## Issue reference

#2342 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by youngbupark** [comment]
linter complains the deprecation of jsonpb. I will fix it by tomorrow.

**Comment by youngbupark** [comment]
@yaron2 #2684 PR needs to merged first to resolve linter errors.

**Comment by youngbupark** [comment]
@yaron2 I realized that I can ignore jsonpb deprecation linter error without #2684 change. Do you want to ignore this check ? or do you want to do the right way after merging #2684 ?

**Comment by yaron2** [comment]
> @yaron2 I realized that I can ignore jsonpb deprecation linter error without #2684 change. Do you want to ignore this check ? or do you want to do the right way after merging #2684 ?

Let's do it the right way :)

**Comment by youngbupark** [comment]
> > @yaron2 I realized that I can ignore jsonpb deprecation linter error without #2684 change. Do you want to ignore this check ? or do you want to do the right way after merging #2684 ?
> 
> Let's do it the right way :)

Then please review https://github.com/dapr/dapr/pull/2684

**Comment by yaron2** [comment]
@youngbupark there's a linter issue?

**Comment by youngbupark** [comment]
> @youngbupark there's a linter issue?

I need to switch from jsonpb (deprecated) to new pkg. I will be able to do by tomorrow morning.

**Comment by youngbupark** [comment]
> > @youngbupark there's a linter issue?
> 
> I need to switch from jsonpb (deprecated) to new pkg. I will be able to do by tomorrow morning.

@yaron2 please review it after build is done.

**Comment by yaron2** [comment]
@youngbupark the build is failing

**Comment by youngbupark** [comment]
> @youngbupark the build is failing

I will fix it tonight

File: pkg/http/api.go [hunk 1]
```diff
@@ -769,9 +769,17 @@ func (a *api) onDirectMessage(reqCtx *fasthttp.RequestCtx) {
...
 		statusCode = invokev1.HTTPStatusFromCode(codes.Code(statusCode))
+		if statusCode != fasthttp.StatusOK {
+			body, err = invokev1.ProtobufToJSON(resp.Status())
+		}
 	}
 
-	respond(reqCtx, statusCode, body)
+	if err != nil {
+		msg := NewErrorResponse("ERR_MALFORMED_REQUEST", err.Error())
+		respondWithError(reqCtx, fasthttp.StatusBadRequest, msg)
```
Comment (rynowak) [inline pkg/http/api.go:779]: The way this is structured is a little confusing because the only case that can cause `err != nil` is line 773.

Comment (rynowak) [reply pkg/http/api.go:779]: Is this really the right status? The "malformed request" in this case would be a failure to serialize `resp.Status()` as JSON.

**Review by rynowak — COMMENTED** [review]
So as I understand this change - in the HTTP -> gRPC invocation path, we now have better reporting of errors coming from gRPC.

The status will be serialized as JSON for the caller to read. Is that the major change here?

It looks like there's another change to better handle propagating of the content type when the callee response with protobuf.

Should we have an E2E test for this? I'm a little surprised we don't have any E2E tests for the error handling code path.

File: pkg/http/api.go (hunk 1 above)
Comment (youngbupark) [reply pkg/http/api.go:779]: You're right. it is to handle L773. I will restructure it. Status code must be Internal Error. I will change it.

**Comment by yaron2** [comment]
General question: In the case of HTTP to HTTP, does this PR change how status codes are handled? in other words, for HTTP to HTTP we should not make decisions based on the callee status code and just proxy what came from the other side.

**Comment by youngbupark** [comment]
> General question: In the case of HTTP to HTTP, does this PR change how status codes are handled? in other words, for HTTP to HTTP we should not make decisions based on the callee status code and just proxy what came from the other side.

HTTP to HTTP, GRPC to GRPC we bypass headers, body, status code without any change.

**Comment by youngbupark** [comment]
> we now have better reporting of errors coming from gRPC.
> The status will be serialized as JSON for the caller to read. Is that the major change here?
YES right

> It looks like there's another change to better handle propagating of the content type when the callee response with protobuf.
If callee returns custom protobuf format, we bypass marshaled protobuf message to caller. 

> Should we have an E2E test for this? I'm a little surprised we don't have any E2E tests for the error handling code path.

@ryanvolum we have no negative e2e tests for invocation. I remember that it was the part of e2e test improvement effort for v1.0. I was thinking that unit-test can cover the current scneario.

I can add the e2e tests for covering all changes over the weekend.

**Comment by yaron2** [comment]
@youngbupark the e2e test can happen next week, really doesn't need to happen on the weekend.

File: pkg/http/api.go (hunk 1 above)
Comment (youngbupark) [reply pkg/http/api.go:779]: I was copying the existing code :) I will commit the changes.

**Comment by rynowak** [comment]
I pulled your changes and tested the HTTP->gRPC path locally, it's working well, and I can get the status/message on the client-side for error cases.

**Comment by youngbupark** [comment]
> I pulled your changes and tested the HTTP->gRPC path locally, it's working well, and I can get the status/message on the client-side for error cases.

👍  I am able to work on the changes based on your comments tonight.

File: pkg/http/api.go (hunk 1 above)
Comment (youngbupark) [reply pkg/http/api.go]: @ryanvolum Please review it again.

Comment (yaron2) [reply pkg/http/api.go]: You mean @rynowak

**Review by yaron2 — APPROVED** [review]
lgtm

**Pull request merged**